toolchain/lowering/lowering_context.cpp

```diff
@@ -94,18 +110,18 @@ auto LoweringContext::BuildFunctionDeclaration(SemIR::FunctionId function_id)
       function_type, llvm::Function::ExternalLinkage,
       semantics_ir().GetString(function.name_id), llvm_module());
 
-  if (has_return_slot) {
-    auto* return_slot = llvm_function->getArg(0);
-    return_slot->addAttr(llvm::Attribute::getWithStructRetType(
-        llvm_context(), GetType(function.return_type_id)));
-    return_slot->setName("return");
-  }
-
-  // Set parameter names.
-  for (auto [i, param_ref] : llvm::enumerate(param_refs)) {
-    auto name_id = semantics_ir().GetNode(param_ref).GetAsParameter();
-    llvm_function->getArg(first_param + i)
-        ->setName(semantics_ir().GetString(name_id));
+  // Set up parameters.
+  for (auto [i, node_id, arg] : llvm::enumerate(
```

Comment:
I was thinking after the first round, is param_nodes efficient here?

IIUC, you're using it to rapidly iterate through the params a second time, versus mirroring the logic on 77-98 to again only take actions conditionally. TBH, I'm not sure what's better: repeating those branches, or having a vector allocation (but also still branches in this loop).

I'm not sure if you have a strong instinct about the best approach here, but maybe it's worth a TODO to see if removing param_nodes and doing a similar param_index as is done in BuildFunctionDefinition would have an impact on performance? Or maybe similarly, storing param_nodes with the Function for use by BuildFunctionDefinition?